enterprise/server/remote_execution/containers/firecracker/firecracker_test.go

```diff
@@ -1655,6 +1661,120 @@ func TestFirecrackerRunWithDockerOverTCPDisabled(t *testing.T) {
 	assert.NotEqual(t, 0, res.ExitCode)
 }
 
+func TestFirecrackerRunWithDockerMirror(t *testing.T) {
+	if *skipDockerTests {
+		t.Skip()
+	}
+
+	tests := []struct {
+		name                 string
+		return404            bool
+		expectedRequestCount int32
+	}{
+		{
+			name:                 "explicitly_pull_from_mirror",
+			return404:            false,
+			expectedRequestCount: 6,
+		},
+		{
+			name:                 "pull_busybox_through_mirror",
+			return404:            false,
+			expectedRequestCount: 6,
+		},
+		{
+			name:                 "mirror_404_fallback",
+			return404:            true,
+			expectedRequestCount: 3,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			flags.Set(t, "executor.task_allowed_private_ips", []string{"default"})
+
+			ctx := context.Background()
+			env := getTestEnv(ctx, t, envOpts{})
+			rootDir := testfs.MakeTempDir(t)
+			workDir := testfs.MakeDirAll(t, rootDir, "work")
+
+			hostIP, err := networking.DefaultIP(ctx)
+			require.NoError(t, err)
+
+			ocireg, err := ociregistry.New(env)
+			require.NoError(t, err)
+
+			port := testport.FindFree(t)
+			listenAddr := fmt.Sprintf("%s:%d", hostIP, port)
+			registryHost := fmt.Sprintf("%s:%d", hostIP, port)
+			registryURL := fmt.Sprintf("http://%s", registryHost)
+
+			var mirrorCounter atomic.Int32
+			handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				log.Infof("Received registry request: %s %s", r.Method, r.URL.Path)
+				mirrorCounter.Add(1)
+				if tc.return404 && strings.HasPrefix(r.URL.Path, "/v2/library/") {
+					w.WriteHeader(http.StatusNotFound)
+					return
+				}
+				ocireg.ServeHTTP(w, r)
+			})
+
+			server := &http.Server{Handler: handler}
+			lis, err := net.Listen("tcp", listenAddr)
+			require.NoError(t, err)
+			go func() { _ = server.Serve(lis) }()
+			t.Cleanup(func() {
+				server.Shutdown(context.Background())
+			})
+
+			opts := firecracker.ContainerOpts{
+				ContainerImage:         imageWithDockerInstalled,
+				ActionWorkingDirectory: workDir,
+				VMConfiguration: &fcpb.VMConfiguration{
+					NumCpus:           1,
+					MemSizeMb:         2500,
+					EnableNetworking:  true,
+					InitDockerd:       true,
+					ScratchDiskSizeMb: 100,
+				},
+				ExecutorConfig: getExecutorConfig(t),
+			}
+			c, err := firecracker.NewContainer(ctx, env, &repb.ExecutionTask{}, opts)
+			require.NoError(t, err)
+
+			configTemplate := `{
+    "insecure-registries": ["%s"],
+    "registry-mirrors": ["%s"]
+}`
+			dockerConfig := fmt.Sprintf(configTemplate, registryHost, registryURL)
+			bashScript := fmt.Sprintf(`set -e
+cat > /etc/docker/daemon.json <<EOF
+%s
+EOF
+kill -HUP $(cat /var/run/docker.pid)  # Reload docker config
+docker pull busybox`, dockerConfig)
```

Comment:
nit: maybe add a TODO here so we don't forget to remove the manual config reloading once we have a way to configure the registry automatically?